### ISSUE TITLE
How to swap existing or insert a new element?

### ISSUE THREAD

**Issue by yopoyka** [bigskysoftware/htmx#2166]
## **The problem:**
I have an element with id task-list.
```html
<div id="task-list" hx-ext="ws" ws-connect="/updates">
    <div id="task-(unique id)"></div>
    ...
</div>
```
It contains a list of elements representing tasks. Each of them has a unique id.
Task-list connects to a websocket endpoint and listens for updates.
Each update consists of html that wraps a task in order to swap it correctly.
```html
<div hx-swap-oob="beforeend:#task-list">
    <div id="task-(unique id)"></div>
</div> 
```
It works for additions and deletions of tasks but not for updates.

## **I'm trying to do the following:**
If task with id exists in the list swap it
Else add the new task to the list

I don't see a way to achieve this behavior without extending htxm.

**Comment by itepastra** [comment]
if you want to update an existing task, you can send the new task like 
`<div id="task-(unique id)" hx-swap-oob="outerHTML"></div>`
instead of the wrapped task.
It will swap the task inline.

**Comment by yopoyka** [comment]
Thanks. But it doesn't solve the problem when task doesn't exist.

**Comment by ChrisLR** [comment]
Any luck finding a solution?
The closest I got so far is to include have a hx-swap-oob delete div preceding the  hx-swap-oob  beforeend div.
It works but the replaced element will be moved.
I would expect beforeend to replace the appropriate id first and only append if its not found.

**Comment by yopoyka** [comment]
No. I've came up with the same solution as yours. You can try to sort items using js on the client if it fits the application.
I've also tried to play with extension api as it allows you to extend swap mechanism but I did not have much time to explore it more. I suggest you to look into it.

**Comment by andryyy** [comment]
Hey, have you tried [idiomorph](https://github.com/bigskysoftware/idiomorph?tab=readme-ov-file#htmx)?

If it works, you may be happy with HTMX v2 coming up. Iirc it will inherit some of the morph magic from it.

**Comment by jurriaan** [comment]
I've had a similar issue where I wanted to prepend elements to a list (using sse instead of websockets).

Hacked my way around the issue using a small htmx extension:

```javascript
let internalApi = null;
htmx.defineExtension('oob-if-exists', {
    init: function (api) {
        internalApi = api;
    },
    transformResponse: function (text, xhr, elt) {
        const fragment = internalApi.makeFragment(text);

        const swapAttr = elt.getAttribute('hx-swap');
        if (swapAttr == 'afterbegin' || swapAttr == 'beforeend') {
            const elements = htmx.findAll(fragment, "[hx-swap-oob=if-exists]");

            for (const element of elements) {
                const selector = '#' + element.id;
                const existingElement = htmx.find(selector);

                if (!!existingElement) {
                    // Move element to the desired spot and swap the HTML
                    existingElement.parentNode.insertAdjacentElement(swapAttr, existingElement);
                    element.setAttribute('hx-swap-oob', 'innerHTML');
                } else {
                    // Just create the element
                    element.removeAttribute('hx-swap-oob');
                }
            }
        }

        const htmlContent = [].map.call(fragment.childNodes, x => x.outerHTML).join('');
        return htmlContent;
    }
});
```

This is the source of the page:
```html
<ol class="task-list" hx-ext="oob-if-exists" sse-swap="update-task" hx-swap="afterbegin">
</ol>
```

With something like this for the tasks:
```javascript
<li id="task-1" hx-swap-oob="if-exists">Test</li>
```

This extension removes the `hx-swap-oob` attribute if its id isn't in the document yet. This will make sure the entries are created if they don't yet exist, but swapped when they are.

It's a bit hacky, but seems to work fine for my use-case and can probably be adapted for yours.

**Comment by schungx** [comment]
What you need is really a binary swap target specification... If selector matches anything do one swap, otherwise do another swap.

So at the very least you'd need two selectors to specify both branches. Currently htmx doesn't seem to allow for that.

Maybe something to be added likr `hx-alt-target` that triggers when the main target returns nothing...

**Comment by LifeLiveOn** [comment]
Hey!, here is how I did it, if adding items with hx-swap="beforeend"

you could use this javascript code to do remove the first appear child or the last depend on your usecase
`<script>
    function handleNewItem() {
        const pantryItems = document.getElementById('pantry-items');
        const lastChild = pantryItems.lastElementChild;
        if (lastChild) {
            const elements = document.querySelectorAll(`#${lastChild.id}`);
            if (elements.length > 1){
                elements[0].parentNode.removeChild(elements[0]);
            }
        }
    }
</script>`

**Comment by maemigh** [comment]
> I've had a similar issue where I wanted to prepend elements to a list (using sse instead of websockets).
> 
> Hacked my way around the issue using a small htmx extension:
> 
> let internalApi = null;
> htmx.defineExtension('oob-if-exists', {
>     init: function (api) {
>         internalApi = api;
>     },
>     transformResponse: function (text, xhr, elt) {
>         const fragment = internalApi.makeFragment(text);
> 
>         const swapAttr = elt.getAttribute('hx-swap');
>         if (swapAttr == 'afterbegin' || swapAttr == 'beforeend') {
>             const elements = htmx.findAll(fragment, "[hx-swap-oob=if-exists]");
> 
>             for (const element of elements) {
>                 const selector = '#' + element.id;
>                 const existingElement = htmx.find(selector);
> 
>                 if (!!existingElement) {
>                     // Move element to the desired spot and swap the HTML
>                     existingElement.parentNode.insertAdjacentElement(swapAttr, existingElement);
>                     element.setAttribute('hx-swap-oob', 'innerHTML');
>                 } else {
>                     // Just create the element
>                     element.removeAttribute('hx-swap-oob');
>                 }
>             }
>         }
> 
>         const htmlContent = [].map.call(fragment.childNodes, x => x.outerHTML).join('');
>         return htmlContent;
>     }
> });
> This is the source of the page:
> 
> <ol class="task-list" hx-ext="oob-if-exists" sse-swap="update-task" hx-swap="afterbegin">
> </ol>
> With something like this for the tasks:
> 
> <li id="task-1" hx-swap-oob="if-exists">Test</li>
> This extension removes the `hx-swap-oob` attribute if its id isn't in the document yet. This will make sure the entries are created if they don't yet exist, but swapped when they are.
> 
> It's a bit hacky, but seems to work fine for my use-case and can probably be adapted for yours.

Exactly what I was looking for

**Comment by C0DK** [comment]
>Maybe something to be added like hx-alt-target that triggers when the main target returns nothing...

I feel like it might be very specific for oob (i might be wrong though) - where you need a `hx-swap-oob-fallback` or something.

I have a similar case where I have a SSE stream either updating or appending to a table. My solution was to track the rows server side. I initially swapped the whole table, but that had annoying consequences with edit fields in the table refreshing as well.

**Comment by C0DK** [comment]
I've run into this issue a bunch of times now, and ended up creating a small extension. I should probably package it into an actual extension, but for now this works fine. It is event driven, (which seems to be the htmx-way). I am not sure how to fix the firefox errors, still getting logged, (the `e.preventDefault();` did not do it).

This was my solution:

```html
<script>
htmx.on('htmx:oobErrorNoTarget', (e) => {
    const elt = e.detail.content;
    const fallback = elt.getAttribute('hx-swap-oob-fallback')
    if (!!fallback) {
        e.preventDefault();
        const [swapStyle, target] = fallback.includes(":") ? fallback.split(":") : ["innerHTML", fallback];

        htmx.swap(target, elt.outerHTML, {swapStyle: swapStyle});
    }
}, {capture: true});
</script>
```

**Comment by 1cg** [comment]
old & answered-ish
